src/qos/kernel.js

```diff
@@ -100,9 +107,10 @@ class QosKernel {
 
     if (!this._cpuLimit) {
       const bucketRange = BUCKET_CEILING - BUCKET_FLOOR
-      const adjustedPercentage = (Game.cpu.bucket - (10000 - bucketRange)) / bucketRange
-      const cpuPercentage = CPU_MINIMUM + ((1 - CPU_MINIMUM) * adjustedPercentage)
-      this._cpuLimit = (Game.cpu.limit * (1 - CPU_ADJUST)) * cpuPercentage
+      const depthInRange = (Game.cpu.bucket - BUCKET_FLOOR) / bucketRange
+      const minToAllocate = Game.cpu.limit * CPU_MINIMUM
+      const maxToAllocate = Game.cpu.tickLimit - CPU_BUFFER
```

Comment:
This should be `const maxToAllocate =  Game.cpu.limit`, since the system doesn't go above `Game.cpu.limit` unless `Game.cpu.bucket > BUCKET_CEILING`, at which point it bursts to keep from wasting CPU.

Reply:
That logic resulted in extremely bursty behaviour in my testing, and I switched to this, which produced a much smoother gradient.

Reply:
I'm okay with choppy behavior when the system has an abundance of bucket as that also means it isn't using all of it's CPU every tick anyways. However, when there isn't an abundance it should scale back (which this does). As long as we maintain low limits on how much each program should typically use the choppiness also shouldn't be much of an issue.

Also keep in mind that this code will never be reached when bucket is create than BUCKET_CEILING, as this logic is below that. So this wouldn't prevent the choppiness anyways.

If you want to change the existing behavior with regards to the BUCKET_CEILING I think it should be a separate pull request (so I can downvote it will still having the option to upvote this one).

Reply:
This smoothly transitions from the "bursting" point to the minimum, rather than a slow linear climb followed by a massive burst. The choppiness is somewhat calmed.